saturn_client/core.py

```diff
@@ -143,6 +170,169 @@ def create_project(
             raise HTTPError(response.status_code, response.json()["message"]) from err
         return response.json()
 
+    def update_project(
+        self,
+        project_id: str,
+        description: Optional[str] = None,
+        image_uri: Optional[str] = None,
+        start_script: Optional[str] = None,
+        environment_variables: Optional[Dict] = None,
+        working_dir: Optional[str] = None,
+        workspace_settings: Optional[Dict] = None,
+        update_jupyter_server: Optional[bool] = True,
+    ) -> Dict[str, Any]:
+        """
+        Create a project from scratch
+
+        :param project_id: ID of project. This is the only field that is required.
+        :param description: Short description of the project (less than 250 characters).
+        :param image_uri: Location of the image. Example:
+            485185227295.dkr.ecr.us-east-1.amazonaws.com/saturn-dask:2020.12.01.21.10
+        :param start_script: Script that runs on start up. Examples: "pip install dask"
+        :param environment_variables: Env vars expressed as a dict. The names will be
+            coerced to uppercase.
+        :param working_dir: Location to use as working directory. Example: /home/jovyan/project
+        :param workspace_settings: Setting for the jupyter associated with the project.
```

Comment:
I think it is easier for people to compose these things using dicts, but I don't mind allowing both approaches.

Reply:
To be clear:

1. I'm not saying the dataclass thing is "another thing people might prefer", but something I actually think is better and should be the main pattern we encourage.
2. I'm not saying these functions should take in either a dict or an instance of one of these classes. I'm saying they should ONLY take / return instances of these classes, and people who really want dictionaries can `.from_dict()` to produce inputs and can `.to_dict()` the outputs.

If you / the team disagree with those points, we should just stick to dictionaries so that at least there's only one way things are done in the client. I wouldn't want to see us support different patterns across different methods, or to have all the methods take in dictionaries and treat the dataclass thing as something optional you can do instead.

Reply:
Right I understand. I am personally pro a dict-only input. But interested in hearing other opinions.

Reply:
Let's see if anyone else has a strong opinion: @saturncloud/saturn-developers

Reply:
Definitely +1 for us supporting and advocating only one way to do things. I don't feel _super_ strongly in either direction, but I do have a preference for the pythonic, here, since this is a client library and not a CLI. Bear in mind that the majority of our initial usage is likely to be from short scripts run from CI - an example of what object vs dict usage there:

```python
import os
from saturn import SaturnConnection

conn = SaturnConnection(os.getenv("SATURN_URL"), os.getenv("SATURN_TOKEN"))

# objects
project = [p for p in conn.list_projects() if p.name == "my-project"][0]
project.update(image=os.getenv("NEW_IMAGE"))

# dicts
project_id = [p for p in conn.list_projects() if p["name"] == "my-project"][0]["id"]
conn.update_project(project_id, image=os.getenv("NEW_IMAGE"))
```

From that usage, they're both pretty similar - two lines, not much going on. I find the object-based version easier to read, but the dict-based version doesn't have to pass the `conn` around.

boto3 is kind of a shining example of a case of dict-based usage gone wrong, and I worry about us going down that path: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/iam-example-server-certificates.html#id3 - boto3 is basically never pleasant to work with.

Stripe, however, is an API that people constantly say is great to work with - but it's _also_ dict-based for its python wrapper: https://stripe.com/docs/api/customers/retrieve - so it may just be a question of documentation.

A happy medium might be returning instances of https://docs.python.org/3/library/collections.html#collections.UserDict (or lists of it) whenever a dict would be returned - that would let the above samples become:

```python
project_id = [p for p in conn.list_projects() if p.name == "my-project"][0].id
conn.update_project(project_id, image=os.getenv("NEW_IMAGE"))
```

That would mean that we don't need to implement and maintain tons of boilerplate, while still making the code not super awful to work with.

Reply:
Sorry, after rereading, I'm not sure that comment actually _says_ anything - I'm not sure that switching everything over to fully class-based is worth the maintenance effort/burden. Having the client be a glorified `curl` wrapper as it is now would not be abnormal nor unexpected for the situation.

Reply:
I am happy to follow the stripe pattern if we can agree that we like that one: https://stripe.com/docs/api. 

I don't feel very strongly about what the responses should be, what I don't want is people to have to think about instantiating classes just to pass these in as kwargs. Specifically I am thinking about `WorkspaceSettings`. I think it is just another level of things to think about for the user and it doesn't make their lives easier.

Reply:
I am going to flatten out `WorkspaceSettings` like you proposed in the original comment @jameslamb :)

Reply:
> switching everything over to fully class-based is worth the maintenance effort/burden

I think it's unfair to characterize my proposal as "switching everything over", since this project's first commit was 4 days ago and it's never been released to any Saturn user.

> Stripe....also dict-based for its python wrapper

More specifically, that client returns dictionaries, but for creating things they have one dedicated class per resource known to that service (e.g. `Charge`, `Customer`). https://github.com/stripe/stripe-python/blob/master/stripe/object_classes.py

> I don't want is people to have to think about instantiating classes just to pass these in as kwargs

I'm unconvinced by this, thanks to the many hours of my life that have been wasted chasing down deeply-nested KeyErrors in `boto3` code. I've also written clients like this for services before, and have been burned by forgetting to add validation code in all of the places that take in such dictionaries as input or return it as output (something you don't have to do passing around instances of dataclasses).

<hr>

Anyway, thanks for hearing me out. Sounds like I'm in the minority, so I can rest my case and let's just move on with dictionaries-only for inputs and outputs. You can revert the `WorkspaceSettings` stuff as well. If we're going to use dicts, let's do it everywhere.